tljh/jupyterhub_config.py

```diff
@@ -13,12 +13,16 @@
 from tljh.yaml import yaml
 from jupyterhub_traefik_proxy import TraefikTomlProxy
 
+from traitlets import Any
+
 class UserCreatingSpawner(SystemdSpawner):
     """
     SystemdSpawner with user creation on spawn.
 
     FIXME: Remove this somehow?
     """
+    user_groups = Any(config=True)
```

Comment:
This should be a `List` traitlet not an `Any`

Reply:
I don't think it can be a List :(
Because *extra_user_groups* will look something like this:
```
{
        'users': {
            'extra_user_groups': {
                "group1": ["u1", "u2"],
                "group2": ["u3", "u4"]
            },
        ....
        }
}
```
and we define ``c.UserCreatingSpawner.user_groups = users['extra_user_groups']``,
*user_groups* will be a dict like:
```
{
    "group1": ["user1", "user2"],
    "group2": ["user3", "user4"]
}
```
Unfortunately, I'm not sure there is a way to add to *UserCreatingSpawner.user_groups* only the list of groups that *user.name* must be added to :disappointed:.

I added some new commits where I defined user_groups to be a Dict traitlet instead of the generic Any.
Do you believe it's possible to have *UserCreatingSpawner* hold only the list of groups for that user and not the entire config, or is Dict good enough?

Reply:
Ah, whoops. I missed this completely, apologies!

I think we shouldn't (right now at least) allow whitelisting like this, since we might not know all the user names at time of config (for most use cases, where the authenticator isn't FirstUseAuthenticator).

Instead, I think user_groups should be a list that's applied to *all* users. This satisfies the use case in #349, which is to help users get access to GPUs, etc. We can work on the per-user case when someone raises that for a specific use cases.

I hope that makes sense, and I'm so sorry it took me this long to respond!

Reply:
It makes perfectly sense. Thanks for taking the time to clarify this and don't worry about the delay :) Let me know if there's anything else I should change.